includes/class-wc-swedbank-pay-invoice-fee.php

```diff
@@ -0,0 +1,198 @@
+
+		add_action( 'wp_ajax_sb_invoice_apply_fee', array( $this, 'ajax_sb_invoice_apply_fee' ) );
+		add_action( 'wp_ajax_nopriv_sb_invoice_apply_fee', array( $this, 'ajax_sb_invoice_apply_fee' ) );
+
+		add_action( 'wp_ajax_sb_invoice_unset_fee', array( $this, 'ajax_sb_invoice_unset_fee' ) );
+		add_action( 'wp_ajax_nopriv_sb_invoice_unset_fee', array( $this, 'ajax_sb_invoice_unset_fee' ) );
+
+		// Payment fee
+		add_action( 'woocommerce_cart_calculate_fees', array( $this, 'add_cart_fee' ) );
+	}
+
+	/**
+	 * WooCommerce Init.
+	 */
+	public function woocommerce_init() {
+		if ( session_status() === PHP_SESSION_NONE && ! headers_sent() ) {
+			session_start();
+		}
+	}
+
+	/**
+	 * WooCommerce Init.
+	 */
+	public function register_post_type() {
+		add_filter(
+			'woocommerce_settings_api_form_fields_payex_checkout',
+			array(
+				$this,
+				'add_settings',
+			)
+		);
+	}
+
+	/**
+	 * Add settings.
+	 *
+	 * @param $form_fields
+	 *
+	 * @return array
+	 */
+	public function add_settings( $form_fields ) {
+		$form_fields['fee'] = array(
+			'title'             => __( 'Invoice Fee', 'swedbank-pay-woocommerce-checkout' ),
+			'type'              => 'number',
+			'custom_attributes' => array(
+				'step' => 'any'
+			),
+			'description'       => __( 'Invoice fee. Set 0 to disable.', 'swedbank-pay-woocommerce-checkout' ),
+			'default'           => '0'
+		);
+
+		$form_fields['fee_is_taxable'] = array(
+			'title'   => __( 'Invoice Fee is Taxable', 'swedbank-pay-woocommerce-checkout' ),
+			'type'    => 'checkbox',
+			'label'   => __( 'Invoice Fee is Taxable', 'swedbank-pay-woocommerce-checkout' ),
+			'default' => 'no'
+		);
+
+		$form_fields['fee_tax_class'] = array(
+			'title'       => __( 'Tax class of Invoice Fee', 'swedbank-pay-woocommerce-checkout' ),
+			'type'        => 'select',
+			'options'     => $this->getTaxClasses(),
+			'description' => __( 'Tax class of Invoice Fee.', 'swedbank-pay-woocommerce-checkout' ),
+			'default'     => 'standard'
+		);
+
+		return $form_fields;
+	}
+
+	/**
+	 * Add fee when selected payment method.
+	 */
+	public function add_cart_fee() {
+		$this->log( __METHOD__ . ': woocommerce_cart_calculate_fees' );
+
+		if ( is_admin() && ! defined( 'DOING_AJAX' ) ) {
+			return;
+		}
+
+		// Is Fee is not specified
+		if ( abs( SB_INVOICE_FEE ) < 0.01 ) {
+			return;
+		}
+
+		$flag = isset( $_SESSION['sb_invoice_fee_apply'] ) ? $_SESSION['sb_invoice_fee_apply'] : null;
+		$this->log( __METHOD__ . var_export( $flag, true ) );
+		if ( 'yes' === $flag ) {
+			$this->log( __METHOD__ . ': Add cart fee.' );
+
+			// Add Fee
+			WC()->cart->add_fee(
+				__( 'Invoice Fee', 'swedbank-pay-woocommerce-checkout' ),
+				SB_INVOICE_FEE,
+				( SB_INVOICE_FEE_IS_TAXABLE === 'yes' ),
+				SB_INVOICE_FEE_TAX_CLASS
+			);
+		}
+	}
+
+	/**
+	 * Ajax: Add invoice fee.
+	 */
+	public function ajax_sb_invoice_apply_fee() {
+		$this->log( __METHOD__ . ': Set invoice fee.' );
+
+		$_SESSION['sb_invoice_fee_apply'] = 'yes';
```

Comment:
Is it impossible to store this information alongside the order? Storing state in `$_SESSION` is a very bad practice.

Reply:
@asbjornu There's order haven't placed yet